File: python/zaga_payments.py
"""Zaga Payments client for Python backends. Standard library only.

    from zaga_payments import ZagaPayments, verify_webhook_signature

    payments = ZagaPayments("https://payments.zaga.co.tz", "zp_live_...")
    charge = payments.create_charge(method="mpesa", amount_minor=5000, currency="TZS", msisdn="+255712345678", idempotency_key=order_id)
    settled = payments.wait_for_settlement(charge["id"])

    # in your webhook view (raw body!)
    ok = verify_webhook_signature(secret, request.headers.get("x-zaga-timestamp"), request.headers.get("x-zaga-signature"), raw_body)
"""
from __future__ import annotations

import hashlib
import hmac
import json
import time
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Callable, Optional


class ZagaPaymentsError(Exception):
    def __init__(self, status: int, code: str, message: str):
        super().__init__(f"{status} {code}: {message}")
        self.status = status
        self.code = code
        self.message = message


class ZagaPayments:
    def __init__(self, base_url: str, api_key: str, timeout: float = 20.0):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout

    def _request(self, method: str, path: str, body: Optional[dict] = None, headers: Optional[dict] = None) -> dict:
        data = None if body is None else json.dumps(body).encode()
        req = urllib.request.Request(self.base_url + path, data=data, method=method)
        req.add_header("authorization", f"Bearer {self.api_key}")
        req.add_header("content-type", "application/json")
        req.add_header("accept", "application/json")
        for k, v in (headers or {}).items():
            req.add_header(k, v)
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                text = resp.read().decode()
                status = resp.status
        except urllib.error.HTTPError as e:
            text = e.read().decode()
            status = e.code
        parsed = json.loads(text) if text else {}
        if not parsed.get("success"):
            raise ZagaPaymentsError(status, parsed.get("code", "ERROR"), parsed.get("message", "Request failed"))
        return parsed

    def list_methods(self, platform: Optional[str] = None) -> list[dict]:
        q = f"?platform={platform}" if platform else ""
        return self._request("GET", f"/v1/methods{q}")["data"]["methods"]

    def list_products(self) -> list[dict]:
        return self._request("GET", "/v1/products")["data"]["products"]

    def create_charge(self, method: str, amount_minor: Optional[int] = None, currency: Optional[str] = None, product_code: Optional[str] = None, msisdn: Optional[str] = None, email: Optional[str] = None, name: Optional[str] = None, description: Optional[str] = None, metadata: Optional[dict] = None, external_reference: Optional[str] = None, return_url: Optional[str] = None, idempotency_key: Optional[str] = None) -> dict:
        body: dict[str, Any] = {"method": method, "customer": {}}
        if amount_minor is not None: body["amount_minor"] = amount_minor
        if currency: body["currency"] = currency
        if product_code: body["product_code"] = product_code
        if msisdn: body["customer"]["msisdn"] = msisdn
        if email: body["customer"]["email"] = email
        if name: body["customer"]["name"] = name
        if description: body["description"] = description
        if metadata: body["metadata"] = metadata
        if external_reference: body["external_reference"] = external_reference
        if return_url: body["return_url"] = return_url
        headers = {"idempotency-key": idempotency_key} if idempotency_key else None
        return self._request("POST", "/v1/charges", body, headers)["data"]

    def get_charge(self, id_or_reference: str, sync: bool = False) -> dict:
        q = "?sync=1" if sync else ""
        return self._request("GET", f"/v1/charges/{urllib.parse.quote(id_or_reference, safe='')}{q}")["data"]

    def wait_for_settlement(self, id_or_reference: str, interval: float = 4.0, timeout: float = 180.0, sync_every: int = 5, on_poll: Optional[Callable[[dict, int], None]] = None) -> Optional[dict]:
        deadline = time.time() + timeout
        attempt = 0
        last = None
        while time.time() < deadline:
            attempt += 1
            charge = self.get_charge(id_or_reference, sync=attempt % sync_every == 0)
            last = charge
            if on_poll: on_poll(charge, attempt)
            if charge["status"] != "pending": return charge
            time.sleep(interval)
        return last

    def refund(self, charge_id: str, amount_minor: Optional[int] = None, reason: Optional[str] = None) -> dict:
        body: dict[str, Any] = {}
        if amount_minor is not None: body["amount_minor"] = amount_minor
        if reason: body["reason"] = reason
        return self._request("POST", f"/v1/charges/{urllib.parse.quote(charge_id, safe='')}/refunds", body)["data"]

    def claim_store_transaction(self, provider: str, transaction_id: str, product_code: Optional[str] = None) -> dict:
        body = {"provider": provider, "transaction_id": transaction_id}
        if product_code: body["product_code"] = product_code
        return self._request("POST", "/v1/store/claims", body)["data"]


def verify_webhook_signature(secret: str, timestamp: Optional[str], signature: Optional[str], body: str | bytes, tolerance_sec: int = 300) -> bool:
    if not timestamp or not signature: return False
    try:
        ts = int(timestamp)
    except ValueError:
        return False
    if abs(time.time() - ts) > tolerance_sec: return False
    raw = body.decode() if isinstance(body, bytes) else body
    expected = hmac.new(secret.encode(), f"{timestamp}.{raw}".encode(), hashlib.sha256).hexdigest()
    received = signature[3:] if signature.startswith("v1=") else signature
    return hmac.compare_digest(expected, received)

File: dart/lib/zaga_payments.dart
/// Zaga Payments client for Dart / Flutter backends and server-side Dart.
///
/// Never ship a live API key inside a mobile app: call the gateway from your
/// backend. This package exists for Dart servers (dart_frog, shelf) and for
/// Flutter apps that only use test keys during development.
///
/// ```dart
/// final payments = ZagaPayments(baseUrl: 'https://payments.zaga.co.tz', apiKey: 'zp_live_…');
/// final charge = await payments.createCharge(method: 'mpesa', amountMinor: 5000, currency: 'TZS', msisdn: '+255712345678', idempotencyKey: orderId);
/// final settled = await payments.waitForSettlement(charge['id']);
/// ```
library zaga_payments;

import 'dart:convert';
import 'dart:io';

import 'package:crypto/crypto.dart';

class ZagaPaymentsException implements Exception {
  final int status;
  final String code;
  final String message;
  ZagaPaymentsException(this.status, this.code, this.message);
  @override
  String toString() => 'ZagaPaymentsException($status $code: $message)';
}

class ZagaPayments {
  final String baseUrl;
  final String apiKey;
  final HttpClient _client = HttpClient();

  ZagaPayments({required String baseUrl, required this.apiKey}) : baseUrl = baseUrl.replaceAll(RegExp(r'/+$'), '');

  Future<Map<String, dynamic>> _request(String method, String path, {Map<String, dynamic>? body, Map<String, String>? headers}) async {
    final request = await _client.openUrl(method, Uri.parse('$baseUrl$path'));
    request.headers.set('authorization', 'Bearer $apiKey');
    request.headers.set('content-type', 'application/json');
    request.headers.set('accept', 'application/json');
    headers?.forEach(request.headers.set);
    if (body != null) request.write(jsonEncode(body));
    final response = await request.close();
    final text = await response.transform(utf8.decoder).join();
    final Map<String, dynamic> parsed = text.isEmpty ? {} : jsonDecode(text) as Map<String, dynamic>;
    if (parsed['success'] != true) {
      throw ZagaPaymentsException(response.statusCode, (parsed['code'] ?? 'ERROR').toString(), (parsed['message'] ?? 'Request failed').toString());
    }
    return parsed;
  }

  /// Methods the buyer can pick right now.
  Future<List<Map<String, dynamic>>> listMethods({String? platform}) async {
    final r = await _request('GET', '/v1/methods${platform == null ? '' : '?platform=$platform'}');
    return List<Map<String, dynamic>>.from(r['data']['methods'] as List);
  }

  Future<List<Map<String, dynamic>>> listProducts() async {
    final r = await _request('GET', '/v1/products');
    return List<Map<String, dynamic>>.from(r['data']['products'] as List);
  }

  /// Starts a payment. Pass [idempotencyKey] (your order id) so retries never double-charge.
  Future<Map<String, dynamic>> createCharge({
    required String method,
    int? amountMinor,
    String? currency,
    String? productCode,
    String? msisdn,
    String? email,
    String? name,
    String? description,
    Map<String, dynamic>? metadata,
    String? externalReference,
    String? returnUrl,
    String? idempotencyKey,
  }) async {
    final r = await _request('POST', '/v1/charges', body: {
      'method': method,
      if (amountMinor != null) 'amount_minor': amountMinor,
      if (currency != null) 'currency': currency,
      if (productCode != null) 'product_code': productCode,
      'customer': {if (msisdn != null) 'msisdn': msisdn, if (email != null) 'email': email, if (name != null) 'name': name},
      if (description != null) 'description': description,
      if (metadata != null) 'metadata': metadata,
      if (externalReference != null) 'external_reference': externalReference,
      if (returnUrl != null) 'return_url': returnUrl,
    }, headers: idempotencyKey == null ? null : {'idempotency-key': idempotencyKey});
    return r['data'] as Map<String, dynamic>;
  }

  /// By id, provider reference or your external reference. [sync] asks the provider directly.
  Future<Map<String, dynamic>> getCharge(String idOrReference, {bool sync = false}) async {
    final r = await _request('GET', '/v1/charges/${Uri.encodeComponent(idOrReference)}${sync ? '?sync=1' : ''}');
    return r['data'] as Map<String, dynamic>;
  }

  /// Polls until the charge leaves `pending` or [timeout] passes. Asks the provider every [syncEvery] polls.
  Future<Map<String, dynamic>?> waitForSettlement(String idOrReference, {Duration interval = const Duration(seconds: 4), Duration timeout = const Duration(minutes: 3), int syncEvery = 5, void Function(Map<String, dynamic> charge, int attempt)? onPoll}) async {
    final deadline = DateTime.now().add(timeout);
    var attempt = 0;
    Map<String, dynamic>? last;
    while (DateTime.now().isBefore(deadline)) {
      attempt++;
      final charge = await getCharge(idOrReference, sync: attempt % syncEvery == 0);
      last = charge;
      onPoll?.call(charge, attempt);
      if (charge['status'] != 'pending') return charge;
      await Future<void>.delayed(interval);
    }
    return last;
  }

  Future<Map<String, dynamic>> refund(String chargeId, {int? amountMinor, String? reason}) async {
    final r = await _request('POST', '/v1/charges/${Uri.encodeComponent(chargeId)}/refunds', body: {if (amountMinor != null) 'amount_minor': amountMinor, if (reason != null) 'reason': reason});
    return r['data'] as Map<String, dynamic>;
  }

  /// Apple/Google in-app purchase, verified server-side and recorded once.
  Future<Map<String, dynamic>> claimStoreTransaction({required String provider, required String transactionId, String? productCode}) async {
    final r = await _request('POST', '/v1/store/claims', body: {'provider': provider, 'transaction_id': transactionId, if (productCode != null) 'product_code': productCode});
    return r['data'] as Map<String, dynamic>;
  }

  void close() => _client.close();
}

/// Verifies a webhook the gateway sent to your backend. Use the raw body bytes.
bool verifyWebhookSignature({required String secret, required String? timestamp, required String? signature, required String body, int toleranceSec = 300}) {
  if (timestamp == null || signature == null) return false;
  final ts = int.tryParse(timestamp);
  if (ts == null) return false;
  if ((DateTime.now().millisecondsSinceEpoch ~/ 1000 - ts).abs() > toleranceSec) return false;
  final expected = _hmacSha256Hex(secret, '$timestamp.$body');
  final received = signature.replaceFirst(RegExp(r'^v1='), '');
  if (expected.length != received.length) return false;
  var diff = 0;
  for (var i = 0; i < expected.length; i++) {
    diff |= expected.codeUnitAt(i) ^ received.codeUnitAt(i);
  }
  return diff == 0;
}

String _hmacSha256Hex(String key, String message) => Hmac(sha256, utf8.encode(key)).convert(utf8.encode(message)).toString();

/// Exposed for tests and for callers that sign their own requests; same primitive the gateway uses.
String hmacSha256HexForTest(String key, String message) => _hmacSha256Hex(key, message);
